### DISCUSSION TITLE
Thredds server return me a extreme big value among multiple thredds environment

### DISCUSSION THREAD

**Discussion by kensinzl** [Unidata/netcdf-java#1226] [General]
We are using the Thredds server 5.4 docker image(unidata/thredds-docker:5.4) and deploying it as a docker container. 

Now I have faced one issue and makes me annoying how to fix

- I have a Java App which is using multiple threads to fetch several NetCDF files from the above Thredds server to boostup the response time. 

- These requests are among a defined lat&lon range, which means each request with different lat&lon value but with the same time range. 

- Majority of response is normal, but some random NetCDF files' response would give me an extremely big value than expected.

- But if I make a single HTTP request for the same params, the extremely big value will back to normal.

eg:
https://<HOST>thredds/ncss/grid/NetCDF_1?var=Var_1&latitude=-37&longitude=174&time_start=2023-08-13T12%3A00%3A00Z&time_end=2023-08-19T12%3A00%3A00Z&timeStride=&vertCoord=&accept=csv
https://<HOST>thredds/ncss/grid/NetCDF_1?var=Var_1&latitude=-36&longitude=176&time_start=2023-08-13T12%3A00%3A00Z&time_end=2023-08-19T12%3A00%3A00Z&timeStride=&vertCoord=&accept=csv
https://<HOST>thredds/ncss/grid/NetCDF_2?var=Var_2&latitude=-37&longitude=174&time_start=2023-08-13T12%3A00%3A00Z&time_end=2023-08-19T12%3A00%3A00Z&timeStride=&vertCoord=&accept=csv
https://<HOST>thredds/ncss/grid/NetCDF_2?var=Var_2&latitude=-36&longitude=176&time_start=2023-08-13T12%3A00%3A00Z&time_end=2023-08-19T12%3A00%3A00Z&timeStride=&vertCoord=&accept=csv
......

if I found the following request response at log showing an extremely big value among multiple threads, then I hit the exactly same HTTP request again by itself, and then the value will be back to normal.

https://<HOST>thredds/ncss/grid/NetCDF_2?var=Var_2&latitude=-36&longitude=176&time_start=2023-08-13T12%3A00%3A00Z&time_end=2023-08-19T12%3A00%3A00Z&timeStride=&vertCoord=&accept=csv

Any idea, please

**Comment by haileyajohnson** [comment]
Is there a reason you need each lat/lon as a separate request? We do have an API for subsetting the data on a single time and a range of lat/lons.

**Reply by kensinzl** [reply]
@haileyajohnson  Thanks for the reply.

Is there a reason you need each lat/lon as a separate request?
=> Yes, we have done this logic for a while, so now still hope to keep this style.


We do have an API for subsetting the data on a single time and a range of lat/lons.
=> 
1. I read the document. The WMS feature has a range of lat/lons with time. 
2. if you did not mention the WMS feature, could you please share one document for your statement, then I can have a look.
3. back to the original question, I still want to know why Thredds server gave an extremely big value via the above request among multiple threads, but can not replicate via a single HTTP request.

Thank you for your help.

**Reply by mnlerman** [reply]
Hello, just to make sure we're on the same page, can you clarify what kind of requests you're making? The original discussion post has NCSS requests but then the last message mentions WMS. 

You can request a [lat/lon bounding box using NCSS](https://docs.unidata.ucar.edu/tds/current/userguide/ncss_grid.html#spatial-and-coordinate-subsetting) which will return a range of lats and lons. netcdf3 and netcdf4-classic are the only supported formats for this type of request 

To answer your question about getting extremely large values, the underlying functionality isn't thread-safe which is why you can't duplicate the problem when you run one request at a time.

**Reply by kensinzl** [reply]
@mnlerman 
thanks for the reply, let us ignore the WMS because when I saw @haileyajohnson message mentioning the range lat/lon, then I remember WMS has this feature and thanks for your document link.


To answer your question about getting extremely large values, the underlying functionality isn't thread-safe which is why you can't duplicate the problem when you run one request at a time.
=> We deployed the TDS among a container(server A), and there is another server(server B) to send the request for the TDS container(server A).

so the multiple threads of requests are coming from server B to server A and also each request is for one NetCDF file with different lat/lon and different time range. so I think it is not directly multiple threads for Threads and should be okay?

**Reply by mnlerman** [reply]
Okay, thanks for the clarification! I just wanted to make sure that we were talking about the same thing. So, even though Server B is requesting multiple files, the data is stored in one file on Server A so each request is reading from that same file, which is were the thread-safety comes in to play.

**Reply by kensinzl** [reply]
@mnlerman 

Thanks for the reply.

even though Server B is requesting multiple files, the data is stored in one file on Server A so each request is reading from that same file

=>I think the scenario should be 
1. Server B send multiple requests to fetch multiple NetCDF files from server A.
2. Each request does fetch one NetCDF file but there are multiple requests so fetch multiple NetCDF files.
3. If you look at my original comment, there is the request example, which only makes the NetCDF file name, var and lat&lon value change among each request.

**Reply by mnlerman** [reply]
Yes, I see the examples on the original post. These two requests:
> https://thredds/ncss/grid/NetCDF_1?var=Var_1&latitude=-37&longitude=174&time_start=2023-08-13T12%3A00%3A00Z&time_end=2023-08-19T12%3A00%3A00Z&timeStride=&vertCoord=&accept=csv

> https://thredds/ncss/grid/NetCDF_1?var=Var_1&latitude=-36&longitude=176&time_start=2023-08-13T12%3A00%3A00Z&time_end=2023-08-19T12%3A00%3A00Z&timeStride=&vertCoord=&accept=csv

 are reading from the same file (NetCDF_1), which, when you make the requests concurrently, will cause problems. I would suggest running your requests synchronously and using a lat/lon bounding box mentioned in my previous comment. If those changes don't resolve the issue you're seeing, please let us know so that we can investigate further but regarding the current scenario, the software is functioning as expected.

**Reply by kensinzl** [reply]
@mnlerman 

Thanks for the reply.

- for the concurrent requests, it seems to be a known issue, and other people have already raised it, eg: https://github.com/Unidata/netcdf-java/pull/1122. 
I am also waiting for the Thredds server image 5.5 at the docker hub to officially be out because this version seems to fix the race condition issue.
If we do multiple requests syn, the response time is not good if faces thousands of requests.

- anyway, back to the lat/lon bounding box request, thanks for the info, but current business logic needs the grid point feature.
even with the bounding box, the Thredds(ServerA) may concurrently receive two requests for the same NetCDF file fetching from Server B and Server C, something like this.

**Reply by mnlerman** [reply]
We are working to get the official release of thredds-5.5 out. Meanwhile, we do have [thredds-5.5-SNAPSHOT available on Docker Hub](https://hub.docker.com/layers/unidata/thredds-docker/5.5-SNAPSHOT/images/sha256-a7fe61a4ac1970e482802394fe7d3cc669788f7f9bf76a1aeac428f815f75cd0?context=explore) which does have the changes that you're looking at. Have you had a chance to test that out? 

Another possible solution you might try is to accessing multiple files asynchronously but staggering points in the same file so that only one request is being sent per file at a time. This would still allow you to get the grid as point features asynchronously without trying to read the same file when it's already open

**Reply by kensinzl** [reply]
@mnlerman 

Thank you for the reply

- Yes, I am aware that there is a 5.5-SNAPSHOT version, but consider it is SNAPSHOT, so better for me to wait until the official version is launched because I use Thredds in the prod environment, just in case.

- could you please share more details for your second sentence for the solution, I want to make sure that I am understanding your meaning for the correct page.
1. accessing multiple files asynchronously but staggering points in the same file so that only one request is being sent per file at a time
=> eg, I have 10 requests from Server B to Thredds(ServerA), these 10 requests are fetching 5 NetCDF files with the same time range and every 2 requests for the same NetCDF file but with different point(lat/lon)

do you mean I should send requests order by NetCDF files rather than without order?

=> without trying to read the same file when it's already open

do you mean when Thredds receive a request, it will open the related NetCDF file, if the next coming request among a 'Thredds configured time range' is still the same NetCDF file, the Thredds will not open the file again? if so, I guess this is the reason you hope for me to order the request by NetCDF name if my understanding is correct for your first sentence.

**Reply by mnlerman** [reply]
I mean the you can make two requests to two separate files asynchronously, but you can't make two requests to one file asynchronously

From the example requests that you posted, I mean that these two requests can run at the same time:
> https://thredds/ncss/grid/NetCDF_1?var=Var_1&latitude=-37&longitude=174&time_start=2023-08-13T12%3A00%3A00Z&time_end=2023-08-19T12%3A00%3A00Z&timeStride=&vertCoord=&accept=csv
> https://thredds/ncss/grid/NetCDF_2?var=Var_2&latitude=-37&longitude=174&time_start=2023-08-13T12%3A00%3A00Z&time_end=2023-08-19T12%3A00%3A00Z&timeStride=&vertCoord=&accept=csv

And these two requests can run at the same time:
> https://thredds/ncss/grid/NetCDF_1?var=Var_1&latitude=-36&longitude=176&time_start=2023-08-13T12%3A00%3A00Z&time_end=2023-08-19T12%3A00%3A00Z&timeStride=&vertCoord=&accept=csv
> https://thredds/ncss/grid/NetCDF_2?var=Var_2&latitude=-36&longitude=176&time_start=2023-08-13T12%3A00%3A00Z&time_end=2023-08-19T12%3A00%3A00Z&timeStride=&vertCoord=&accept=csv

because they are not reading from the same file. I think you're encountering problems when you're reading NetCDF_1 on thread 1 and then thread 2 tries to read NetCDF_1 and resets the cursor. If you wait for the response from NetCDF_1, then send the next request for NetCDF_1, you shouldn't encounter the same problems. One way to achieve this might be to send all of the requests for for a file on a single thread so that there can't be multiple concurrent requests to the same file.

**Reply by kensinzl** [reply]
@mnlerman 

Thanks for the details information.

em.... seems if the Thredds can not deal with multiple requests concurrently well for one NetCDF, that will be inconvenient here.

So far, there are two isolated servers to hit the Thredds server.

Considering that more and more users come to use it in the future, it is hard to avoid there being no two requests for the same NetCDF file concurrently. Meanwhile if do the syn logic, the performance will be slow for business logic which needs a huge amount of requests.

**Comment by haileyajohnson** [comment]
Hi @kensinzl , yes I expect you are right that the issue is fixed by #1122
What @mnlerman was trying to offer was just a workaround to minimize the chance of hitting the race condition you're seeing. The ideal solution would be to use the 5.5-SNAPSHOT. We are working to get the official release out, but at the moment the latest SNAPSHOT release IS the recommended version to be used in production (5.4 has a few security exploits that are fixed in 5.5-SNAPSHOT. It would also help us if you could just test out 5.5 and let us know is the issue is resolved.

Thanks!

**Comment by kensinzl** [comment]
@mnlerman @mnlerman 

Thanks for the reply. Now I will do the following several things to have a try.

- try the 5.5-SNAPSHOT at the UAT version
- Considering this issue may bring from the concurrent scenario, I will have a re-try HTTP request when I I receive the extremely big value
- last thing, I had a quick try at the range of lan/lat for each request, and it seems did not give me the JSON response, and it auto downloads the subset(smaller) NetCDF file. I may miss something. if possible, could you please share a URL, then I can have a hit.

This is my current grid point URL, then it works
<Host>/thredds/ncss/grid/<NetCDF File>?var=uv02&latitude=-37&longitude=174&time_start=2023-08-17T18%3A00%3A00Z&time_end=2023-08-19T18%3A00%3A00Z&timeStride=&vertCoord=&accept=csv

I changed it with the North, West, East and South, but not work? any idea
<Host>/thredds/ncss/grid/<NetCDF File>?var=uv02&north=-33.941&west=166.093&east=179.703&south=-47.677&time_start=2023-08-17T18%3A00%3A00Z&time_end=2023-08-19T18%3A00%3A00Z&timeStride=&vertCoord=&accept=csv

Thanks so much, hope you also have a great weekend. thanks this whole week help.

**Reply by mnlerman** [reply]
Hello,

> I changed it with the North, West, East and South, but not work? any idea
> /thredds/ncss/grid/?var=uv02&north=-33.941&west=166.093&east=179.703&south=-47.677&time_start=2023-08-17T18%3A00%3A00Z&time_end=2023-08-19T18%3A00%3A00Z&timeStride=&vertCoord=&accept=csv

the CSV format is only compatible with singe point locations. you can change `accept=csv` to `accept=netcdf3` in the request

**Reply by kensinzl** [reply]
@mnlerman 
Thanks for the reply. I list what I have tried, and please have a look.

/thredds/ncss/grid/<NetCDF_File>?var=uv02&north=-33.941north=-33.941&west=166.093&east=179.703&south=-47.677&horizStride=1&time_start=2023-08-21T12:00:00Z&time_end=2023-08-21T13:00:00Z&&&accept=netcdf3

=> Tried the above URL for TDS from csv into netcdf3, and it made me download a subset(smaller) NetCDF file, but I hope it can give me the JSON(CSV) response like Grid Single Point because now my client needs the JSON(CSV) file, something like this pattern "2023-02-13T12:00:00Z, GridPointRequestedAt[37.000S_174.000E],-36.996,174.000,291.5084"

Thank you so much

**Reply by kensinzl** [reply]
@mnlerman @haileyajohnson  Hey team, could someone please have a look at the above message when you feel available, thanks so much.

**Reply by kensinzl** [reply]
@haileyajohnson @mnlerman 

One more thing, I have started to use the 5.5-snapshot of TDS docker version. 

Now I have found one issue with the DAP4 API, its URLs
https://HOST_NAME/thredds/dap4/Model_NAME/NetCDF.nc.dmr.xml

I also attached the exception track stack,

If I degrade to 5.4, the same NetCDF file for the above URL will work.

Thank you so much


> thredds-local  | java.lang.UnsupportedOperationException: DAP4 is not currently functional, but we are working on it!
thredds-local  | 	at thredds.server.dap4.Dap4Controller.handleRequest(Dap4Controller.java:68) ~[classes/:5.5-SNAPSHOT]
thredds-local  | 	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
thredds-local  | 	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:?]
thredds-local  | 	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
thredds-local  | 	at java.lang.reflect.Method.invoke(Method.java:566) ~[?:?]
thredds-local  | 	at org.springframework.web.method.support.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:205) ~[spring-web-5.3.27.jar:5.3.27]
thredds-local  | 	at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:150) ~[spring-web-5.3.27.jar:5.3.27]
thredds-local  | 	at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:117) ~[spring-webmvc-5.3.27.jar:5.3.27]
thredds-local  | 	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:895) ~[spring-webmvc-5.3.27.jar:5.3.27]
thredds-local  | 	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:808) ~[spring-webmvc-5.3.27.jar:5.3.27]
thredds-local  | 	at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:87) ~[spring-webmvc-5.3.27.jar:5.3.27]
thredds-local  | 	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1072) [spring-webmvc-5.3.27.jar:5.3.27]
thredds-local  | 	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:965) [spring-webmvc-5.3.27.jar:5.3.27]
thredds-local  | 	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1006) [spring-webmvc-5.3.27.jar:5.3.27]
thredds-local  | 	at org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:898) [spring-webmvc-5.3.27.jar:5.3.27]
thredds-local  | 	at javax.servlet.http.HttpServlet.service(HttpServlet.java:489) [servlet-api.jar:?]
thredds-local  | 	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:883) [spring-webmvc-5.3.27.jar:5.3.27]
thredds-local  | 	at javax.servlet.http.HttpServlet.service(HttpServlet.java:583) [servlet-api.jar:?]
thredds-local  | 	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:212) [catalina.jar:8.5.92]
thredds-local  | 	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:156) [catalina.jar:8.5.92]
thredds-local  | 	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:51) [tomcat-websocket.jar:8.5.92]
thredds-local  | 	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:181) [catalina.jar:8.5.92]
thredds-local  | 	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:156) [catalina.jar:8.5.92]
thredds-local  | 	at org.apache.catalina.filters.CorsFilter.handleNonCORS(CorsFilter.java:330) [catalina.jar:8.5.92]
thredds-local  | 	at org.apache.catalina.filters.CorsFilter.doFilter(CorsFilter.java:155) [catalina.jar:8.5.92]
thredds-local  | 	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:181) [catalina.jar:8.5.92]
thredds-local  | 	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:156) [catalina.jar:8.5.92]
thredds-local  | 	at org.apache.catalina.filters.HttpHeaderSecurityFilter.doFilter(HttpHeaderSecurityFilter.java:129) [catalina.jar:8.5.92]
thredds-local  | 	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:181) [catalina.jar:8.5.92]
thredds-local  | 	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:156) [catalina.jar:8.5.92]
thredds-local  | 	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:337) [spring-security-web-5.7.10.jar:5.7.10]
thredds-local  | 	at thredds.servlet.filter.RequestBracketingLogMessageFilter.doFilter(RequestBracketingLogMessageFilter.java:50) [classes/:5.5-SNAPSHOT]
thredds-local  | 	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:346) [spring-security-web-5.7.10.jar:5.7.10]
thredds-local  | 	at thredds.servlet.filter.RequestQueryFilter.doFilter(RequestQueryFilter.java:90) [classes/:5.5-SNAPSHOT]
thredds-local  | 	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:346) [spring-security-web-5.7.10.jar:5.7.10]
thredds-local  | 	at thredds.servlet.filter.HttpHeadFilter.doFilter(HttpHeadFilter.java:47) [classes/:5.5-SNAPSHOT]
thredds-local  | 	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:346) [spring-security-web-5.7.10.jar:5.7.10]
thredds-local  | 	at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:221) [spring-security-web-5.7.10.jar:5.7.10]
thredds-local  | 	at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:186) [spring-security-web-5.7.10.jar:5.7.10]
thredds-local  | 	at org.springframework.web.filter.DelegatingFilterProxy.invokeDelegate(DelegatingFilterProxy.java:354) [spring-web-5.3.27.jar:5.3.27]
thredds-local  | 	at org.springframework.web.filter.DelegatingFilterProxy.doFilter(DelegatingFilterProxy.java:267) [spring-web-5.3.27.jar:5.3.27]
thredds-local  | 	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:181) [catalina.jar:8.5.92]
thredds-local  | 	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:156) [catalina.jar:8.5.92]
thredds-local  | 	at org.apache.logging.log4j.web.Log4jServletFilter.doFilter(Log4jServletFilter.java:71) [log4j-web-2.17.1.jar:2.17.1]
thredds-local  | 	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:181) [catalina.jar:8.5.92]
thredds-local  | 	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:156) [catalina.jar:8.5.92]
thredds-local  | 	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:167) [catalina.jar:8.5.92]
thredds-local  | 	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:90) [catalina.jar:8.5.92]
thredds-local  | 	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:483) [catalina.jar:8.5.92]
thredds-local  | 	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:130) [catalina.jar:8.5.92]
thredds-local  | 	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:93) [catalina.jar:8.5.92]
thredds-local  | 	at org.apache.catalina.valves.AbstractAccessLogValve.invoke(AbstractAccessLogValve.java:682) [catalina.jar:8.5.92]
thredds-local  | 	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:74) [catalina.jar:8.5.92]
thredds-local  | 	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:343) [catalina.jar:8.5.92]
thredds-local  | 	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:617) [tomcat-coyote.jar:8.5.92]
thredds-local  | 	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:63) [tomcat-coyote.jar:8.5.92]
thredds-local  | 	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:932) [tomcat-coyote.jar:8.5.92]
thredds-local  | 	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1694) [tomcat-coyote.jar:8.5.92]
thredds-local  | 	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:52) [tomcat-coyote.jar:8.5.92]
thredds-local  | 	at org.apache.tomcat.util.threads.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1191) [tomcat-util.jar:8.5.92]
thredds-local  | 	at org.apache.tomcat.util.threads.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:659) [tomcat-util.jar:8.5.92]
thredds-local  | 	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61) [tomcat-util.jar:8.5.92]
thredds-local  | 	at java.lang.Thread.run(Thread.java:829) [?:?]

**Reply by haileyajohnson** [reply]
Hi, thanks for the update! DAP4 is, in fact, not currently functional, but we _are_ working on it. It was not functional in 5.4 either, it just did not return that execption.

**Reply by kensinzl** [reply]
Hey, @haileyajohnson  for the reply.

Currently we are using this DAP4 API to fetch one attribute from the response(XML style) at the 5.4, which is suitable for us one particular business case.

I just found oh... same URL for 5.5-snapshot is throwing the exception for me, so I made this post to let you know, so your team will fix this issue at the later commit, right?

once again, thank you so much

**Reply by haileyajohnson** [reply]
@kensinzl, we are working on DAP4 support (you can follow along with pull requests such as [this one](https://github.com/Unidata/tds/pull/394)) and we will re-enable DAP4 requests once it is operational.

**Reply by kensinzl** [reply]
@haileyajohnson Thanks for the reply and I will keep an eye on this pull request.

On the other hand, could you please have a look at the previous message, I just copied the following.

Please help me double confirm, from my test, I think the range lat/lon for TDS will let you download a subset NetCDF file rather than give you a json response. Once again, thank you so much

----------------------------------------------------------------------------------------------------------------------------------

Thanks for the reply. I list what I have tried, and please have a look.

/thredds/ncss/grid/<NetCDF_File>?var=uv02&north=-33.941north=-33.941&west=166.093&east=179.703&south=-47.677&horizStride=1&time_start=2023-08-21T12:00:00Z&time_end=2023-08-21T13:00:00Z&&&accept=netcdf3

=> Tried the above URL for TDS from csv into netcdf3, and it made me download a subset(smaller) NetCDF file, but I hope it can give me the JSON(CSV) response like Grid Single Point because now my client needs the JSON(CSV) file, something like this pattern "2023-02-13T12:00:00Z, GridPointRequestedAt[37.000S_174.000E],-36.996,174.000,291.5084"

Thank you so much

**Reply by haileyajohnson** [reply]
That is correct, csv is not supported for grid requests.

**Reply by kensinzl** [reply]
@haileyajohnson 

Thanks for the reply. Yes, I know csv is not supported. 

now I have changed to netcdf3, and the response is to let me download a subset of NetCDF rather than give me the JSON response.

So my question is 1. Does a range lon/lat request behaviour like this, e.g.: download a subset NetCDF file? 2. can a range lon/lat request return a JSON response? 

thank you so much
